### PR TITLE
Fix: Do not cache `.eslintrc.js` (fixes #5067)

### LINKED ISSUE #5067: Do not cache `.eslintrc.js` configs
Currently the `.eslintrc.js` config is cached because it is loaded with `require`. If changes are made to the config while running the same `eslint` process, those changes will not be picked up.
This is an extension of #4610.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30189136-do-not-cache-eslintrc-js-configs?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by MethodGrab**
Read `.eslintrc.js` on every config load by not using `require`
Fixes: #5067
Related: #4610

### CODE REVIEW

File: tests/lib/config/config-file.js [hunk 1]
```diff
@@ -349,6 +365,32 @@ describe("ConfigFile", function() {
             assert.deepEqual(config, updatedConfig.eslintConfig);
         });
 
+        it("should load fresh information from a .eslintrc.js file", function() {
+            var initialConfig = {
```
Comment (nzakas) [inline tests/lib/config/config-file.js]: Won't this leave a temp file hanging around? I don't think you have to overwrite a file to test this functionality.

**Comment by nzakas** [comment]
Please check the build failure

File: tests/lib/config/config-file.js (hunk 1 above)
Comment (MethodGrab) [inline tests/lib/config/config-file.js]: I used the same technique as #4610 ([config-file.js#L317](https://github.com/eslint/eslint/blob/master/tests/lib/config/config-file.js#L317)). What would be the preferred way of doing it? Stubs?

Comment (nzakas) [inline tests/lib/config/config-file.js]: You just need to check that the value is unique. Load the same file twice and use `!==` to ensure they're not the same object.

**Comment by MethodGrab** [comment]
The test failures are caused by a non-existant `eslint-config-foo` module, is this being mocked or something?

File: tests/lib/config/config-file.js (hunk 1 above)
Comment (MethodGrab) [inline tests/lib/config/config-file.js]: Comparison with `!==` works when requiring directly but `ConfigFile.load` always seems to return a new object:

```
it.only("should load fresh information from a .eslintrc.js file v2", function() {
    var config1 = ConfigFile.load(getFixturePath("js/.eslintrc.js"));
    var config2 = ConfigFile.load(getFixturePath("js/.eslintrc.js"));

    var config3 = require(getFixturePath("js/.eslintrc.js"));
    var config4 = require(getFixturePath("js/.eslintrc.js"));

    console.log( 'config1 !== config2', config1 !== config2 ); // --> true
    console.log( 'config3 !== config4', config3 !== config4 ); // --> false
    assert.notStrictEqual(config1, config2, 'via ConfigFile.load');
    assert.notStrictEqual(config3, config4, 'via require');
});
```

**Comment by nzakas** [comment]
Probably. Try renaming?

**Comment by nzakas** [comment]
Oops, I meant rebasing. (Stupid autocorrect)

**Comment by MethodGrab** [comment]
I've just rebased again, it's definitely stubbing related. The 2 failing tests stub out `resolve.sync` ([config-file.js#L165](https://github.com/eslint/eslint/blob/master/tests/lib/config/config-file.js#L165)) and pass the contents of the test config **object** as the `filePath` to `loadJSConfigFile`. Because `requireUncached` is not stubbed, it tries to load the non-existant module rather than just returning the supplied config object.

Whats the best way to resolve this? Stub `requireUncached` as well?

**Comment by nzakas** [comment]
Seems like it.

**Comment by MethodGrab** [comment]
:white_check_mark: The tests are passing now. Not sure if that was the best way to mock it, but it works!

File: tests/lib/config/config-file.js [hunk 2]
```diff
@@ -497,6 +519,32 @@ describe("ConfigFile", function() {
...
+                    globals: {},
+                    rules: {
+                        quotes: 0
+                    }
+                },
+                tmpFilename = ".eslintrc.js",
+                tmpFilePath = writeTempJsConfigFile(initialConfig, tmpFilename),
+                config = ConfigFile.load(tmpFilePath);
+            assert.deepEqual(config, initialConfig);
+            writeTempJsConfigFile(updatedConfig, tmpFilename, path.dirname(tmpFilePath));
```
Comment (ilyavolodin) [inline tests/lib/config/config-file.js]: This would leave temp file hanging around after tests, right? You should probably clean it up.

Comment (MethodGrab) [inline tests/lib/config/config-file.js]: It was using [raszi/node-tmp](https://github.com/raszi/node-tmp), the same as the existing `writeTempConfigFile` function, which is supposed to clean up the files automatically when the process exits.

I've changed it to use [bruce/node-temp](https://github.com/bruce/node-temp) which successfully cleans up the temp dir automatically.

Do you want me to amend this PR/create a new one so `writeTempConfigFile` also cleans up the temp files?

Comment (ilyavolodin) [inline tests/lib/config/config-file.js]: Ahh.. I see. temp.track() is supposed to clean up on exit. That sounds fine to me.

**Comment by ilyavolodin** [comment]
LGTM. Need another set of eyes on this.

**Comment by mysticatea** [comment]
Looks good to me.

**Pull request merged**